### DISCUSSION TITLE
Batches for Transductive Learning

### DISCUSSION THREAD

**Discussion by LuisaWerner** [pyg-team/pytorch_geometric#4097] [Q&A] [answered]
Hello everyone, 

I have a question about sampling batches with NeighborLoader(). So far I have done inductive training on my data graph and sampled batches as follows:

train_loader = NeighborLoader(data,
                                  num_neighbors=[args.sampling_neighbor_size] * 2,
                                  shuffle=True,
                                  input_nodes=split_idx['train'],
                                  batch_size=args.batch_size)

As fas as I understood the NeighborLoader samples batches in a way that the first |batch_size| nodes are the ones I want to make predictions for and the following ones are the necessary neighbours. So I make predictions like this: 

out = model(batch.x, batch.adj_t)[:args.batch_size]

Now, in a transductive learning setting, I would like the nodes for which I make predictions (the first |batch_size| nodes of the batch) to come ONLY from the training set while the sampled neighbour nodes from whatever subset. 
Do you know the best way to realise this? 
In the parameter input_nodes I can only specify in general from which subset nodes are sampled but I cannot differentiate between neighbors and the "main" nodes for which I want to make predictions, right? 

Does anyone know how I can create batches for a transductive learning setting? 

Cheers, 
Luisa :-)

**Comment by wsad1** [comment, marked as answer]
You can use the `batch_size` in the data object returned by the `loader`, the first `batch_size` nodes will be in `input_nodes`.
```
loader = NeighborLoader(data,num_neighbors=[args.sampling_neighbor_size] * 2,
shuffle=True,input_nodes=split_idx['train'],batch_size=args.batch_size)
data_batch = next(iter(loader))
data_batch.x[:data_batch.batch_size] # contains only features of nodes in input_nodes
```

**Reply by LuisaWerner** [reply]
Thanks for your reply! 

But if I understood correctly, this again only creates "inductive" batches where all nodes (the nodes I want to make predictions for as well as the sampled neighbors) come from the train set. 

How can I achieve that the first |batch_size| nodes are from the train set and the sampeld neighbors are from train/valid or test set? 

Does anyone have an idea how to solve this?

**Reply by wsad1** [reply]
Let me know if I am missing something. But with this the nodes you want to make a prediction on will come from the train set and the neighbors can be from train/val or test set.

**Reply by LuisaWerner** [reply]
Thank you for your help! 
Maybe I didn't write this explicitly enough but I would like to have a train_loader object through which I can iterate later on in the training loop. So each batch should be a Data object with the nodes (the first ones of the training set) and an adjacency matrix. 
Or did I get something wrong in general? Does the input_nodes parameter only refer to neighbors or also the "main" nodes for which I want to predict?

**Reply by wsad1** [reply]
The `input_nodes` parameter refers only to the nodes you want to sample, so in your case it only refers to the "main" nodes you want to predict on, and does not refer to neighbors. Let me know if that's not clear.

**Reply by LuisaWerner** [reply]
Ah so that means "by default" the batches are transductive and the neighbors are sampled from the full data object? 
So in this case I would need to remove valid and test nodes from the data object that the sampler returns inductive batches (= where the neighbors only come from the same set as specified in input_nodes). Is this correct? 

It would be amazing to have a little example in the documentation to see how the batches sampled by the neighbor loader are composed 😊 

Thanks a lot for your help!

**Reply by wsad1** [reply]
You'd like only only train nodes in the `Data` object returned by `NeighborLoader`. Is my understanding correct?

If so you could use [Data.subgraph](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Data.subgraph) to get a graph with only train nodes and then run `NeighborLoader`.

**Reply by LuisaWerner** [reply]
Yes, exactly! This answers my question. Thanks again for your help!
